Adding feature to include token generation for long process of spark.

rupesh3020 opened the pull request:
As we have feature to add HTTP headers in the header configuration but this token expires in one hour and spline does not have a mechanism to recreate token and make requests.

This feature adds the functionality by adding a new configuration authentication, where we can specify 
* clientid, clientsecret and other details to get tokens on the fly

Commit:
+ Added new optional authentication argument in Rest Client
+ If this contains client credential grant type variables then use auth
else normal request
+ Added authentication in config for http dispatcher types

Commit:
Merge branch 'develop' into feature/createoptionalauthenticationmechanism

cerveada commented on core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/rest/RestEndpoint.scala:
```diff
@@ -26,14 +26,41 @@ import java.io.ByteArrayOutputStream
 
-  def head(): HttpResponse[String] = request
+  def withAuth(): HttpRequest = {
+    if(authentication.contains("clientId") && authentication.contains("clientSecret") && authentication.contains("tokenUrl") && authentication.contains("scope")) {
```
Those `contains` will always fail, they are call on the option not on the map.

I think the wrapping Option is not needed here, we can just use empty map when no authentication is required.

rupesh3020 replied:
Sure, I have changed from option to a map

rupesh3020 replied:
@wajda @cerveada I have tested it. please let me know if something else is required here

cerveada commented on core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/rest/RestEndpoint.scala:
```diff
@@ -26,14 +26,41 @@ import java.io.ByteArrayOutputStream
+    val jsonResp = scala.util.parsing.json.JSON.parseFull(resp.body)
+    jsonResp match {
+      case Some(map: Map[String, Any]) => map.getOrElse("access_token", "").toString
+      case _ => ""
```
Why not fail here? Without the token, the request will fail anyway, right?

rupesh3020 replied:
Done , I have raised a exception if token is not matched

cerveada commented:
Is this implementing Bearer authentification? (https://datatracker.ietf.org/doc/html/rfc6750)

I think it would make sense to have `authentification.type` property to allow other types in the future.

cerveada commented on core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/rest/RestEndpoint.scala:
```diff
@@ -26,14 +26,41 @@ import java.io.ByteArrayOutputStream
 
   def post(data: String, contentType: String, enableRequestCompression: Boolean): HttpResponse[String] = {
-    val jsonRequest = request
+    val jsonRequest = withAuth()
```
This will request a new token before each request. Isn't it enough to get the token just once?

wajda replied:
@cerveada by "once" you mean at the first method call and cache it? The point here is that we cannot obtain it on the init time because it can expire by the time the job has finished.

cerveada replied:
Well, then it depends on how quickly it expires. Maybe it is not wort it to do the caching.

rupesh3020 replied:
As we know that the request is only made when there is an action. This creates token when post is done. We can also put a function to check the expiry. Let me know if it is ok.

rupesh3020 replied:
Done, now it checks if token is expired or not. If it is expired then it creates token again

cerveada commented:
Currently, the build is failing because test do not compile. The tests must be modified to accommodate the changes in the dispatcher.

Commit:
- Removed Option from Map of authentication
+ Added functionality to check if token expired and only recreate token
if expired

Commit:
Added tokenCache name

rupesh3020 committed:
Merge branch 'AbsaOSS:develop' into develop

rupesh3020 committed:
Added AuthenticationFactory.scala to create auth class as per argument in authentication map

rupesh3020 committed:
Merge pull request #1 from rupesh3020/feature/addauthmodule

Added AuthenticationFactory.scala to create auth class as per argumen…

Commit:
Added NoAuthentication class and fixed tests

Commit:
fixing code smells

Commit:
fixing code smells

Commit:
fixing code smells

rupesh3020 commented:
> Is this implementing Bearer authentification? (https://datatracker.ietf.org/doc/html/rfc6750)
> 
> I think it would make sense to have `authentification.type` property to allow other types in the future.

Commit:
fixing code smells

Commit:
Adding documentation for authentication in README.md

Commit:
Adding documentation for authentication in README.md

Commit:
Adding documentation for authentication in README.md

Commit:
Adding documentation for authentication in README.md

rupesh3020 commented:
> Currently, the build is failing because test do not compile. The tests must be modified to accommodate the changes in the dispatcher.

@cerveada Tests are fixed and i have tested the jar, it works fine. can you please check

cerveada commented:
@rupesh3020 thanks for fixing the tests and other issues, I will take a look.

cerveada commented on core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/rest/RestClient.scala:
```diff
@@ -41,14 +41,16 @@ object RestClient extends Logging {
     logDebug(s"readTimeout = $readTimeout")
     logDebug(s"disableSslValidation = $disableSslValidation")
     logDebug(s"headers = $headers")
+    logInfo(s"authe= $authentication")
```
We should be careful to not log any secrets

cerveada commented on core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/rest/AuthenticationFactory.scala:
```diff
@@ -0,0 +1,102 @@
+
+object AuthenticationFactory{
+  def createAuthentication(authentication: scala.collection.immutable.Map[String, String]): Authentication = {
+    authentication("grantType") match {
```
This will throw an exception when the map is empty. I would check for empty map and if it is empty I would use NoAuthentication

cerveada replied:
for `mode` if it's disabled, just use NoAuthentication

cerveada commented on core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/rest/AuthenticationFactory.scala:
```diff
@@ -0,0 +1,102 @@
+  private val clientSecret: String = authentication.getOrElse("clientSecret","")
+  private val tokenUrl: String = authentication.getOrElse("tokenUrl","")
+  private val grantType: String = authentication.getOrElse("grantType","")
+  private val scope: String = authentication.getOrElse("scope","")
```
For mandatory properties, use just ```authentication("clientId")``` it will fail when the property is missing, but that's what we want if it's mandatory.

cerveada commented on core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/rest/AuthenticationFactory.scala:
```diff
@@ -0,0 +1,102 @@
+  override def createRequest(httpRequest: HttpRequest, authentication: scala.collection.immutable.Map[String, String]): HttpRequest ={
+    authentication.get("mode") match {
+      case Some("enabled") => {
+        if (authentication.contains("clientId") && authentication.contains("clientSecret") && authentication.contains("tokenUrl") && authentication.contains("scope")) {
```
All those checks should be done already (when the class is instantiated) here it should be just get the token and put it in the header.

cerveada commented on core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/rest/AuthenticationFactory.scala:
```diff
@@ -0,0 +1,102 @@
+  private val grantType: String = authentication.getOrElse("grantType","")
+  private val scope: String = authentication.getOrElse("scope","")
+
+  private def isTokenExpired(expirationTime: Instant): Boolean = {
+    Instant.now().isAfter(expirationTime.minusSeconds(300) )
+  }
```
Good idea to extract the expiration checking logic into a separated method. I would go even further and use a `isTokenValid` method where you can check both, that it is defined and the expiration.

Commit:
fixed comments

Commit:
fixed comments

Commit:
fixed comments

rupesh3020 commented:
@cerveada Issues fixed. Can you please check once. I have done functional testing

cerveada commented on core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/rest/AuthenticationFactory.scala:
```diff
@@ -0,0 +1,105 @@
+
+case class ClientCredentialsAuthentication(authentication: scala.collection.immutable.Map[String, String]) extends Authentication with Logging {
+  case class Token(tokenValue: String, expirationTime: Instant)
+  private val tokenCache: scala.collection.mutable.Map[String, Token] = scala.collection.mutable.Map.empty[String, Token]
```
Since you have just one token, no need to use a Map.

Please use `private var tokenCache: Option[Token] = None`

rupesh3020 replied:
var is rejected by code analyser

cerveada commented on core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/rest/AuthenticationFactory.scala:
```diff
@@ -0,0 +1,105 @@
+  private def isTokenInvalid: Boolean = {
+    val cachedToken = tokenCache.get("token")
+    cachedToken.isEmpty || Instant.now().isAfter(cachedToken.getOrElse(throw new IllegalArgumentException(failureMessage)).expirationTime.minusSeconds(300))
+  }
```
Since you already checked `cachedToken.isEmpty` you can do `cachedToken.get` instead of `getOrElse`. there is no need for exception here. Token not being set simply means it is invalid.

rupesh3020 replied:
It fails in static code analysis. Both in using var and using get . Can you suggest ?

cerveada replied:
That is OK, generally using var and get on Option should be avoided, but in this case the cache must be mutable. And calling get on Option is OK if you know that it contains value.

rupesh3020 replied:
@cerveada done.. thanks for comment

cerveada commented on core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/rest/AuthenticationFactory.scala:
```diff
@@ -0,0 +1,105 @@
+    if (authentication.isDefinedAt("mode") && authentication.isDefinedAt("grantType")) {
+      authentication("mode") match {
+          case "enabled" => authentication("grantType") match {
+                                  case "client_credentials" => ClientCredentialsAuthentication(authentication)
+                                  case _ =>
+                                    logInfo(s"$authentication('grantType') not implemented")
+                                    NoAuthentication(authentication)
```
too much indentation

cerveada replied:
When `authentication.mode` is `enabled` and some mandatory parameter is missing, we should throw an exception.

cerveada commented on core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/rest/AuthenticationFactory.scala:
```diff
@@ -0,0 +1,105 @@
+        case Some(map: scala.collection.immutable.Map[String, Any]) =>
+          val token = map.getOrElse("access_token", "").toString
+          if (token.nonEmpty) {
+            val expirationTime = Instant.now().plus(Duration.ofSeconds(map.getOrElse("expires_in", 0).toString.toDouble.toInt))
```
`toString.toDouble.toInt`
why are all those conversions needed?

Commit:
fixed comments

cerveada commented on core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/rest/AuthenticationFactory.scala:
```diff
@@ -0,0 +1,104 @@
+
+      val jsonResp = scala.util.parsing.json.JSON.parseFull(resp.body)
+      jsonResp match {
+        case Some(map: scala.collection.immutable.Map[String, Any]) =>
+          val token = map.getOrElse("access_token", "").toString
+          if (token.nonEmpty) {
+            val expirationTime = Instant.now().plus(Duration.ofSeconds(map.getOrElse("expires_in", 0).toString.toLong))
+            val newToken = Token(token, expirationTime)
+            tokenCache = Some(newToken)
+          }
+          token
```
In this case `access_token` is mandatory in the response(https://www.rfc-editor.org/rfc/rfc6749#section-5.1), so you may just assume it is there.

No need for `map.getOrElse("access_token", "")` or checking `token.nonEmpty`.

Just use `map("access_token")`

rupesh3020 replied:
Done

cerveada commented on core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/rest/AuthenticationFactory.scala:
```diff
@@ -0,0 +1,104 @@
+          throw new RuntimeException(failureMessage)
+      }
+    } else {
+      tokenCache.getOrElse(throw new RuntimeException(failureMessage)).tokenValue
```
in this place `isTokenInvalid` was already checked, so you can safely do `tokenCache.get.tokenValue`

rupesh3020 replied:
@cerveada done

rupesh3020 replied:
@cerveada  Can you please merge if everything looks good as we are waiting for this feature.

Commit:
fixed comments

Commit:
fixed comments

cerveada reviewed (approved):
LGTM

wajda commented:
Sorry guys, I was really buried with other stuff recently, but I really would like to have a quick look it it. Give me an hour or two, I'll check it out. In general, yes, that's awesome feature, I'm not going to delay merge.

rupesh3020 committed:
Merge branch 'develop' into develop

wajda commented:
@rupesh3020 Thank you very much for your contribution!. We've made a few changes to your code and merged it as another pull request - #631. So I'm closing this one for that reason.
